### PR TITLE
Disable Fused Multiply-Add instructions on WCOSS2 when `-DFASTER=ON`

### LINKED ISSUE #1619: Two HAFS forecasts crashed on WCOSS2 when Fused Multiply-Add instructions were enabled
## Description
Two HAFS forecasts crashed with the new `-DFASTER=ON` optimization options, which was tracked down to the use of fused multiply-add (FMA) instructions. Turning them off via `-no-fma` allowed the tests to pass. 

### To Reproduce:
FIXME: Someone moved the Dogwood test data, and mine is on Cactus, which I cannot access.

## Additional context

There is more information in #1618, the pull request that fixes this.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
## Description
Two HAFS-B tests crashed on WCOSS2 with the new `-DFASTER=ON` mode unless fused multiply-add instructions were disabled. This isn't too surprising since the Intel compiler documentation warns that the compiler isn't guaranteed to optimize correctly for non-Intel processors. WCOSS2 uses AMD processors.

To fix this problem, this PR adds `-no-fma` only on WCOSS2, and only when `-DFASTER=ON` is enabled. You can also send `-no-fma` on demand by setting `-DDISABLE_FMA=ON`. Developer @BijuThomas-NOAA has independently verified this fixes the HAFS-B crashes.

The reason the crash only happens with `-DFASTER=ON` is that the default Release optimization already disables FMA through the `-fp-model consistent`. While the vague documentation of `-fp-model consistent` is inconsistent from version to version of the Intel compiler, that part at least is consistent.

### Input data additions/changes
- [x] No changes are expected to input data.

### Anticipated changes to regression tests:
- [x] Changes are expected to the following tests: `hafs_regional_storm_following_1nest_atm_ocn_wav` on WCOSS2 and Acorn *only*.

That is because it is the only test that uses the `-DFASTER=ON` option, and the change is *only on WCOSS2 and Acorn*.

## Subcomponents involved:
- [ ] AQM
- [ ] CDEPS
- [ ] CICE
- [ ] CMEPS
- [ ] CMakeModules
- [ ] FV3
- [ ] GOCART
- [ ] HYCOM
- [ ] MOM6
- [ ] NOAHMP
- [ ] WW3
- [ ] stochastic_physics
- [x] none

## Commit Queue Checklist:

- [ ] Link PR's from all sub-components involved
- [ ] Confirm reviews completed in sub-component PR's
- [ ] Add all appropriate labels to this PR.
- [ ] Run full RT suite on either Hera/Cheyenne with both Intel/GNU compilers
- [ ] Add list of any failed regression tests to "Anticipated changes to regression tests" section.

## Linked PR's and Issues:
- closes #1619

## Testing Day Checklist:

- [ ] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR.
- [ ] Move new/updated input data on RDHPCS Hera and propagate input data changes to all supported systems.

### Testing Log (for CM's):
- RDHPCS
  - Intel
    - [ ] Hera
    - [ ] Orion
    - [ ] Jet
    - [ ] Gaea
    - [ ] Cheyenne
  - GNU
    - [ ] Hera
    - [ ] Cheyenne
- WCOSS2
  - [ ] Dogwood/Cactus
  - [ ] Acorn
- CI
  - [ ] Completed
- opnReqTest
  - [ ] N/A
  - [ ] Log attached to comment

### CODE REVIEW

**Comment by junwang-noaa** [comment]
@SamuelTrahanNOAA Do we want to use -nofma on wcoss2 only for FASTER option or it might be good that we turn off fma on all the platforms?

**Comment by SamuelTrahanNOAA** [comment]
> Do we want to use -nofma on wcoss2 only for FASTER option or it might be good that we turn off fma on all the platforms?

I don't think that is necessary; only turning it of on AMD platforms should be sufficient. Intel's compiler is written only to optimize for Intel processors, and they test it thoroughly on their own processors. Throughout their documentation, you will see warnings that it may not optimize correctly for non-Intel processors. We're just seeing one instance of that.

**Comment by junwang-noaa** [comment]
Thanks, @SamuelTrahanNOAA. Do you see any performance impact on wcoss2 when turning off fma?

**Comment by SamuelTrahanNOAA** [comment]
> Thanks, @SamuelTrahanNOAA. Do you see any performance impact on wcoss2 when turning off fma?

I expect some performance impact, but not a terribly large one, since it's one of many optimizations and instruction sets enabled by FASTER=ON. The HAFS team has not told me how large the impact is, so I'll contact them again for an update on their testing.

File: cmake/Intel.cmake [hunk 1]
```diff
@@ -43,3 +43,9 @@ if(APPLE)
   # Passing the -c flag includes them and fixes an error with undefined symbols
   set(CMAKE_Fortran_ARCHIVE_FINISH "<CMAKE_RANLIB> -c <TARGET>")
 endif()
+
+# This must be last, to override all other optimization settings.
+if(DISABLE_FMA)
+  set(CMAKE_Fortran_FLAGS_RELEASE "${CMAKE_Fortran_FLAGS_RELEASE} -no-fma")
+  set(CMAKE_C_FLAGS_RELEASE "${CMAKE_Fortran_FLAGS_RELEASE} -no-fma")
+endif()
```
Comment (BrianCurtis-NOAA) [inline cmake/Intel.cmake:51]: @SamuelTrahanNOAA Could we include the -no-fma option inside the FASTER section so we don't need to add a special "DISABLE_FMA" in CMakeLists ?

Comment (SamuelTrahanNOAA) [reply cmake/Intel.cmake:51]: No. This is just a workaround for WCOSS2 and Acorn. The platform-specific options are not in Intel.cmake, so there must be another flag to pass -no-fma in.

Comment (DusanJovic-NOAA) [reply cmake/Intel.cmake:51]: We have CMAKE_Platform variable defined. You can use it to identify WCOSS2 and Acorn platforms and set appropriate flags.

Comment (SamuelTrahanNOAA) [reply cmake/Intel.cmake:51]: That's not how other parts of the build work. The design of the build system is for platform-specific settings to be in `cmake/configure_platform.compiler.cmake`. Breaking the design just to eliminate one `-D` option is worse than having the `-D`

Also, that option is *not* only needed for the WCOSS2 and Acorn systems. It is needed for other AMD-based systems as well. It just so happens that WCOSS2 and Acorn have the combination of compiler and chip that cause this to break.

If you hide the workaround from users that may need it, not only are you breaking the design, but you're also making the platform porting experience objectively worse.

**Comment by SamuelTrahanNOAA** [comment]
I'm converting this to a draft so it is not accidentally merged before I finish speed tests with Biju.

**Comment by SamuelTrahanNOAA** [comment]
According to @BijuThomas-NOAA's tests, the difference in runtime, if any, is much less than random variability. For HAFS, at least, there's no disadvantage to this change. This is from his email. Runtimes are in seconds, and the test was done on the production side of WCOSS2.

```
HAFSA
fma    5471.756311
nofma  5479.343784


HAFSB
fma    5430.821674
nofma  5427.427221
```

I apologize for the delay; the HAFS team has been busy with the implementation and have had to prioritize their time, and I don't have production access.

**Comment by BrianCurtis-NOAA** [comment]
I ran this on WCOSS2, and i can confirm only hafs_regional_storm_following_1nest_atm_ocn_wav failed.

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Comment by SamuelTrahanNOAA** [comment]
> I ran this on WCOSS2, and i can confirm only hafs_regional_storm_following_1nest_atm_ocn_wav failed.

Good, good. That is as expected since the results of that test will change.

**Comment by zach1221** [comment]
Hi, @SamuelTrahanNOAA We're going to start working this PR next, would you be able to create another PR to have it merged into #1640 ?

**Comment by SamuelTrahanNOAA** [comment]
> Hi, @SamuelTrahanNOAA We're going to start working this PR next, would you be able to create another PR to have it merged into #1640 ?

I'm willing to have the two merged into a single PR. Could you clarify: are you asking me to open a new PR that combines both sets of changes?

**Comment by SamuelTrahanNOAA** [comment]
I have opened the combined PR here. I've never been asked to do it this way though, so please confirm this is what you wanted.

https://github.com/ufs-community/ufs-weather-model/pull/1645

**Comment by DusanJovic-NOAA** [comment]
Opening yet another PR is completely unnecessary. You could simply merge the changes from #1640 in this PR. Or the #1640 owner could simply pull the changes from this PR.

**Comment by zach1221** [comment]
@DusanJovic-NOAA Sorry for the miscommunication, since 1645 had been created already we'll move forward with that PR.

**Comment by SamuelTrahanNOAA** [comment]
> Opening yet another PR is completely unnecessary. You could simply merge the changes from https://github.com/ufs-community/ufs-weather-model/pull/1640 in this PR. Or the https://github.com/ufs-community/ufs-weather-model/pull/1640 owner could simply pull the changes from this PR.

Indeed, the request confused me too, but since they asked me to open a new PR, I did.

**Pull request closed without merging**